PathWPV2 Plugin installation breaks config

## Issue description

Installtion PathWPv2 together with a bunch of other plugins broke our instance.


### Bug Report

#### Describe how to reproduce the bug

We tried to upgrade from scm-manager 1.60 to scm-manager 2.9 today and occured an error after trying to install various plugins, including Path Write Protection, which leads to an out of bound exception.

#### Describe your SCM-Manager setup

We use 
scm-manager 2.9, 
pathWP 2.0.1
ldap 2.0.1
branchwp 2.0.0
git-plugin 2.9.1
svn-plugin 2.9.1
hg-plugin 2.9.1

#### Which operating system, browser and versions do you use?

The server is running on a debian 10,
we tried to install the plugin using firefox in windows.

#### Log of the bug

```
2020-11-13 15:31:42.243 [main] [          ] INFO  sonia.scm.lifecycle.classloading.ClassLoaderLifeCycle - create new simple ClassLoaderLifeCycle
2020-11-13 15:31:42.248 [main] [          ] INFO  sonia.scm.lifecycle.BootstrapContextFilter - register for restart events
2020-11-13 15:31:42.251 [main] [          ] INFO  sonia.scm.event.LegmanScmEventBus - create new event bus ScmEventBus-1
2020-11-13 15:31:42.304 [main] [          ] INFO  sonia.scm.event.ScmEventBus - use sonia.scm.event.LegmanScmEventBus as event bus implementation
2020-11-13 15:31:42.359 [main] [          ] INFO  sonia.scm.lifecycle.BootstrapContextListener - start scm-manager initialization
2020-11-13 15:31:43.017 [main] [          ] INFO  sonia.scm.plugin.PluginProcessor - collect plugins
2020-11-13 15:31:43.204 [main] [          ] INFO  sonia.scm.plugin.PluginProcessor - install plugin tree:
+- scm-ldap-plugin@2.0.1
+- scm-branchwp-plugin@2.0.0
+- scm-pathwp-plugin@2.0.1
+- scm-git-plugin@2.9.1
+- scm-legacy-plugin@2.9.1
+- scm-hg-plugin@2.9.1
+- scm-svn-plugin@2.9.1

2020-11-13 15:31:44.409 [main] [          ] INFO  sonia.scm.update.repository.XmlRepositoryV1UpdateStep - no v1 repositories database file found
2020-11-13 15:31:44.574 [main] [          ] INFO  sonia.scm.update.UpdateEngine - running update step for type sonia.scm.branchwp.config.repository.xml and version 2.0.0 (class sonia.scm.branchwp.update.BranchWPV2RepositoryConfigMigrationUpdateStep)
2020-11-13 15:31:44.631 [main] [          ] INFO  sonia.scm.update.UpdateEngine - running update step for type sonia.scm.pathwp.config.repository.xml and version 2.0.0 (class sonia.scm.pathwp.update.PathWPV2RepositoryConfigMigrationUpdateStep)
2020-11-13 15:31:44.667 [main] [          ] ERROR sonia.scm.migration.UpdateException - could not execute update for type sonia.scm.pathwp.config.repository.xml to version 2.0.0 in class sonia.scm.pathwp.update.PathWPV2RepositoryConfigMigrationUpdateStep
java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
	at sonia.scm.pathwp.update.PathWPV2RepositoryConfigMigrationUpdateStep.createV2Permission(PathWPV2RepositoryConfigMigrationUpdateStep.java:108)
	at sonia.scm.pathwp.update.PathWPV2RepositoryConfigMigrationUpdateStep.buildConfig(PathWPV2RepositoryConfigMigrationUpdateStep.java:94)
	at sonia.scm.pathwp.update.PathWPV2RepositoryConfigMigrationUpdateStep.lambda$doUpdate$1(PathWPV2RepositoryConfigMigrationUpdateStep.java:74)
	at sonia.scm.update.MapBasedPropertyReaderInstance.call(MapBasedPropertyReaderInstance.java:58)
	at sonia.scm.update.MapBasedPropertyReaderInstance.lambda$forEachEntry$0(MapBasedPropertyReaderInstance.java:44)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.Collections$UnmodifiableMap$UnmodifiableEntrySet.lambda$entryConsumer$0(Collections.java:1576)
	at java.base/java.util.HashMap$EntrySpliterator.forEachRemaining(HashMap.java:1746)
	at java.base/java.util.Collections$UnmodifiableMap$UnmodifiableEntrySet$UnmodifiableEntrySetSpliterator.forEachRemaining(Collections.java:1601)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at sonia.scm.update.MapBasedPropertyReaderInstance.forEachEntry(MapBasedPropertyReaderInstance.java:44)
	at sonia.scm.pathwp.update.PathWPV2RepositoryConfigMigrationUpdateStep.doUpdate(PathWPV2RepositoryConfigMigrationUpdateStep.java:74)
	at sonia.scm.update.UpdateEngine.execute(UpdateEngine.java:87)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at sonia.scm.update.UpdateEngine.update(UpdateEngine.java:77)
	at sonia.scm.lifecycle.BootstrapContextListener.processUpdates(BootstrapContextListener.java:175)
	at sonia.scm.lifecycle.BootstrapContextListener.startEitherMigrationOrApplication(BootstrapContextListener.java:133)
	at sonia.scm.lifecycle.BootstrapContextListener.createMigrationOrNormalModuleProvider(BootstrapContextListener.java:124)
	at sonia.scm.lifecycle.BootstrapContextListener.createStageTwoInjector(BootstrapContextListener.java:117)
	at sonia.scm.lifecycle.BootstrapContextListener.getInjector(BootstrapContextListener.java:93)
	at com.google.inject.servlet.GuiceServletContextListener.contextInitialized(GuiceServletContextListener.java:45)
	at sonia.scm.lifecycle.BootstrapContextListener.contextInitialized(BootstrapContextListener.java:75)
	at sonia.scm.lifecycle.BootstrapContextFilter.initializeContext(BootstrapContextFilter.java:74)
	at sonia.scm.lifecycle.BootstrapContextFilter.init(BootstrapContextFilter.java:65)
	at org.eclipse.jetty.servlet.FilterHolder.initialize(FilterHolder.java:140)
	at org.eclipse.jetty.servlet.ServletHandler.lambda$initialize$0(ServletHandler.java:739)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:734)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at org.eclipse.jetty.servlet.ServletHandler.initialize(ServletHandler.java:763)
	at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:379)
	at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1457)
	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1422)
	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:911)
	at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:288)
	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:524)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:117)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
	at org.eclipse.jetty.server.Server.start(Server.java:423)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:110)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
	at org.eclipse.jetty.server.Server.doStart(Server.java:387)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at sonia.scm.server.ScmServer.init(ScmServer.java:130)
	at sonia.scm.server.ScmServer.run(ScmServer.java:91)
	at sonia.scm.server.ScmServerDaemon.main(ScmServerDaemon.java:53)
2020-11-13 15:31:44.668 [main] [          ] ERROR sonia.scm.lifecycle.BootstrapContextListener - failed to create stage two injector
sonia.scm.migration.UpdateException: could not execute update for type sonia.scm.pathwp.config.repository.xml to version 2.0.0 in class sonia.scm.pathwp.update.PathWPV2RepositoryConfigMigrationUpdateStep
	at sonia.scm.update.UpdateEngine.execute(UpdateEngine.java:90)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at sonia.scm.update.UpdateEngine.update(UpdateEngine.java:77)
	at sonia.scm.lifecycle.BootstrapContextListener.processUpdates(BootstrapContextListener.java:175)
	at sonia.scm.lifecycle.BootstrapContextListener.startEitherMigrationOrApplication(BootstrapContextListener.java:133)
	at sonia.scm.lifecycle.BootstrapContextListener.createMigrationOrNormalModuleProvider(BootstrapContextListener.java:124)
	at sonia.scm.lifecycle.BootstrapContextListener.createStageTwoInjector(BootstrapContextListener.java:117)
	at sonia.scm.lifecycle.BootstrapContextListener.getInjector(BootstrapContextListener.java:93)
	at com.google.inject.servlet.GuiceServletContextListener.contextInitialized(GuiceServletContextListener.java:45)
	at sonia.scm.lifecycle.BootstrapContextListener.contextInitialized(BootstrapContextListener.java:75)
	at sonia.scm.lifecycle.BootstrapContextFilter.initializeContext(BootstrapContextFilter.java:74)
	at sonia.scm.lifecycle.BootstrapContextFilter.init(BootstrapContextFilter.java:65)
	at org.eclipse.jetty.servlet.FilterHolder.initialize(FilterHolder.java:140)
	at org.eclipse.jetty.servlet.ServletHandler.lambda$initialize$0(ServletHandler.java:739)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:734)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at org.eclipse.jetty.servlet.ServletHandler.initialize(ServletHandler.java:763)
	at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:379)
	at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1457)
	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1422)
	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:911)
	at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:288)
	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:524)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:117)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
	at org.eclipse.jetty.server.Server.start(Server.java:423)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:110)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
	at org.eclipse.jetty.server.Server.doStart(Server.java:387)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at sonia.scm.server.ScmServer.init(ScmServer.java:130)
	at sonia.scm.server.ScmServer.run(ScmServer.java:91)
	at sonia.scm.server.ScmServerDaemon.main(ScmServerDaemon.java:53)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
	at sonia.scm.pathwp.update.PathWPV2RepositoryConfigMigrationUpdateStep.createV2Permission(PathWPV2RepositoryConfigMigrationUpdateStep.java:108)
	at sonia.scm.pathwp.update.PathWPV2RepositoryConfigMigrationUpdateStep.buildConfig(PathWPV2RepositoryConfigMigrationUpdateStep.java:94)
	at sonia.scm.pathwp.update.PathWPV2RepositoryConfigMigrationUpdateStep.lambda$doUpdate$1(PathWPV2RepositoryConfigMigrationUpdateStep.java:74)
	at sonia.scm.update.MapBasedPropertyReaderInstance.call(MapBasedPropertyReaderInstance.java:58)
	at sonia.scm.update.MapBasedPropertyReaderInstance.lambda$forEachEntry$0(MapBasedPropertyReaderInstance.java:44)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.Collections$UnmodifiableMap$UnmodifiableEntrySet.lambda$entryConsumer$0(Collections.java:1576)
	at java.base/java.util.HashMap$EntrySpliterator.forEachRemaining(HashMap.java:1746)
	at java.base/java.util.Collections$UnmodifiableMap$UnmodifiableEntrySet$UnmodifiableEntrySetSpliterator.forEachRemaining(Collections.java:1601)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at sonia.scm.update.MapBasedPropertyReaderInstance.forEachEntry(MapBasedPropertyReaderInstance.java:44)
	at sonia.scm.pathwp.update.PathWPV2RepositoryConfigMigrationUpdateStep.doUpdate(PathWPV2RepositoryConfigMigrationUpdateStep.java:74)
	at sonia.scm.update.UpdateEngine.execute(UpdateEngine.java:87)
	... 45 common frames omitted
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,10 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## Unreleased
+###
+- Fixed broken migration with disabled or invalid configuration ([#10](https://github.com/scm-manager/scm-pathwp-plugin/pull/10), [#11](https://github.com/scm-manager/scm-pathwp-plugin/issues/11) and [#12](https://github.com/scm-manager/scm-pathwp-plugin/pull/12))
+
 ## 2.0.01 - 2020-06-17
 ### Added
 - Documentation in English and German ([#6](https://github.com/scm-manager/scm-pathwp-plugin/pull/6))

--- a/src/main/java/sonia/scm/pathwp/update/PathWPV2RepositoryConfigMigrationUpdateStep.java
+++ b/src/main/java/sonia/scm/pathwp/update/PathWPV2RepositoryConfigMigrationUpdateStep.java
@@ -91,7 +91,7 @@ public class PathWPV2RepositoryConfigMigrationUpdateStep implements UpdateStep {
 
     List<PathWritePermission> mappedPermissions = new ArrayList<>();
     for (String v1Permission : splitV1Permissions) {
-      mappedPermissions.add(createV2Permission(v1Permission));
+      createV2Permission(v1Permission).ifPresent(mappedPermissions::add);
     }
 
     PathWritePermissions v2Permissions = new PathWritePermissions();
@@ -101,14 +101,17 @@ public class PathWPV2RepositoryConfigMigrationUpdateStep implements UpdateStep {
     return of(v2Permissions);
   }
 
-  private PathWritePermission createV2Permission(String v1Permission) {
+  private Optional<PathWritePermission> createV2Permission(String v1Permission) {
     String[] splitV1Permission = v1Permission.split(",");
+    if (splitV1Permission.length != 2) {
+      return empty();
+    }
 
     String path = splitV1Permission[0];
     String name = splitV1Permission[1].replaceAll("@","");
     boolean group = splitV1Permission[1].contains("@");
 
-    return new PathWritePermission(path, name, group, Type.ALLOW);
+    return of(new PathWritePermission(path, name, group, Type.ALLOW));
   }
 
   @Override

--- a/src/main/java/sonia/scm/pathwp/update/PathWPV2RepositoryConfigMigrationUpdateStep.java
+++ b/src/main/java/sonia/scm/pathwp/update/PathWPV2RepositoryConfigMigrationUpdateStep.java
@@ -83,7 +83,7 @@ public class PathWPV2RepositoryConfigMigrationUpdateStep implements UpdateStep {
     LOG.debug("migrating repository specific pathwp configuration for repository id {}", repositoryId);
 
     String v1Permissions = properties.get(PATHWP_PERMISSIONS);
-    if (Strings.isNullOrEmpty(v1Permissions)) {
+    if (Strings.isNullOrEmpty(v1Permissions) || v1Permissions.equals("false")) {
       return empty();
     }
 

--- a/src/test/java/sonia/scm/pathwp/update/PathWPV2RepositoryConfigMigrationUpdateStepTest.java
+++ b/src/test/java/sonia/scm/pathwp/update/PathWPV2RepositoryConfigMigrationUpdateStepTest.java
@@ -79,6 +79,19 @@ public class PathWPV2RepositoryConfigMigrationUpdateStepTest {
   }
 
   @Test
+  public void shouldMigrateDisabledSetting() {
+    ImmutableMap<String, String> mockedValues =
+      ImmutableMap.of(
+        "pathwp.permissions", "false"
+      );
+    testUtil.mockRepositoryProperties(new V1PropertyDaoTestUtil.PropertiesForRepository(REPO_NAME, mockedValues));
+
+    updateStep.doUpdate();
+
+    assertThat(getConfigStore().get()).isNull();
+  }
+
+  @Test
   public void shouldSkipRepositoriesIfPermissionsAreEmpty() {
     ImmutableMap<String, String> mockedValues =
       ImmutableMap.of(
@@ -89,6 +102,20 @@ public class PathWPV2RepositoryConfigMigrationUpdateStepTest {
     updateStep.doUpdate();
 
     assertThat(getConfigStore().get()).isNull();
+  }
+
+  @Test
+  public void shouldSkipRepositoriesIfPermissionsAreIncomplete() {
+    ImmutableMap<String, String> mockedValues =
+      ImmutableMap.of(
+        "pathwp.permissions", "[/test/path/,]"
+      );
+    testUtil.mockRepositoryProperties(new V1PropertyDaoTestUtil.PropertiesForRepository(REPO_NAME, mockedValues));
+
+    updateStep.doUpdate();
+
+    assertThat(getConfigStore().get().isEnabled()).isFalse();
+    assertThat(getConfigStore().get().getPermissions()).isEmpty();
   }
 
   private ConfigurationStore<PathWritePermissions> getConfigStore() {

